base/reinterpretarray.jl

```diff
@@ -351,6 +365,10 @@ end
 @inline @propagate_inbounds function _getindex_ra(a::NonReshapedReinterpretArray{T,N,S}, i1::Int, tailinds::TT) where {T,N,S,TT}
     # Make sure to match the scalar reinterpret if that is applicable
     if sizeof(T) == sizeof(S) && (fieldcount(T) + fieldcount(S)) == 0
+        if sizeof(T) == 0 # singleton types
+            @boundscheck checkbounds(a, i1, tailinds...)
+            return T.instance
```

Comment:
Can we add a test to make sure that:

```
mutable struct A
end
```

Which is also `sizeof(A)==0` is not allowed? I was also wondering if `A()` would be better but meh.
Maybe be more cautious and do:
```
sizeof(T) == 0 && isdefined(T, :instance)
```

Reply:
Of course, thank you for the suggestion! I added the tests in the new commit 33a73f80c4.

I also replaced all the `sizeof(T) == 0` conditions by `isdefined(T, :instance)` since that's the practical definition of a singleton type, from https://github.com/JuliaLang/julia/blob/acb6e16b5b941cd3efc7ddec2217a4497120f8bb/base/reflection.jl#L660
But, in any case, both conditions are actually equivalent here because the indirect implication always holds and, in our case, `sizeof(T) == 0` implies `isdefined(T, :instance)` because building a `reinterpretarray` towards type `T` requires it to be `isbitstype` (which precludes mutable types).

By the way I have to use `T.instance` instead of `T()` because the latter may not be defined if the internal constructor is overloaded:
```julia
struct SomeSingleton
    # A singleton type that does not have the internal constructor SomeSingleton()
    SomeSingleton(x) = new()
end
```
I adapted a few tests as well to use this `SomeSingleton` type when doing `getindex` and `setindex` operations.

Reply:
Yes of course, that's much clearer. Done in d49d605583, thanks again for the review!